src/shared/ui/pickerPrompter.ts

```diff
@@ -0,0 +1,255 @@
+                }
+            }
+        })
+        picker.show()
+    }).finally(() => picker.hide())
+}
+
+export class QuickPickPrompter<T> extends Prompter<T> {
+    protected _lastPicked?: DataQuickPickItem<T>
+
+    public set lastResponse(response: DataQuickPickItem<T> | undefined) {
+        if (response === undefined || !isDataQuickPickItem(response)) {
+            return
+        }
+
+        this.quickPick.activeItems = this.quickPick.items.filter(item => item.label === response.label)
+
+        if (this.quickPick.activeItems.length === 0) {
+            this.quickPick.activeItems = [this.quickPick.items[0]]
+        }
+    }
+
+    public get lastResponse() {
+        return this._lastPicked
+    }
+
+    constructor(public readonly quickPick: DataQuickPick<T>) {
+        super()
+    }
+
+    public setSteps(current: number, total: number): void {
+        this.quickPick.step = current
+        this.quickPick.totalSteps = total
+    }
+
+    protected async promptUser(): Promise<PromptResult<T>> {
+        const choices = await promptUser(this.quickPick)
+
+        if (choices === undefined) {
+            return choices
+        }
+
+        this._lastPicked = choices[0]
+        const result = choices[0].data
+
+        return result instanceof Function ? await result() : result
+    }
+
+    public setLastResponse(picked: DataQuickPickItem<T> | undefined = this._lastPicked): void {
+        if (picked === undefined) {
+            return
+        }
+
+        this.quickPick.activeItems = this.quickPick.items.filter(item => item.label === picked.label)
+
+        if (this.quickPick.activeItems.length === 0) {
+            this.quickPick.activeItems = [this.quickPick.items[0]]
+        }
+    }
+}
+
+/**
+ * Allows the prompter to accept the QuickPick filter box as input, shown as a QuickPickItem.
+ * Note: untested with any picker that can refresh items (haven't needed this functionality yet)
+ *
+ * @param label The label of the QuickPickItem that shows the user's input
+ * @param transform Required when the expected type is not a string, transforming the input into the expected type or a control signal.
+ */
+export class CustomQuickPickPrompter<T> extends QuickPickPrompter<T> {
+    constructor(quickPick: DataQuickPick<T>, label: string, transform: (v: string) => PromptResult<T>) {
+        super(quickPick)
+        this.addCustomInput(label, transform)
+    }
+
+    private addCustomInput(label: string, transform: (v: string) => PromptResult<T>): void {
+        const picker = this.quickPick as DataQuickPick<T | symbol>
+        let items: DataQuickPickItem<T | symbol>[] | undefined = [...(picker.items as DataQuickPickItem<T | symbol>[])]
+        let lastUserInput: string | undefined
+
+        function update(value: string = '') {
+            if ((items === undefined || items.length === 0) && picker.busy === false) {
+                items = [...(picker.items as DataQuickPickItem<T | symbol>[])]
```

Comment:
If you're already defining this as an array, why spread operator the array into a new array, and spreading it again later?

Reply:
Remnant of older code. I'll be refactoring this method for clarity.